### PR TITLE
Switch Hashtables to Dictionary in the encoding classes so Hashtable can be linked out.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mcs/class/corlib/ReferenceSources/EncodingTable.cs [hunk 1]
```diff
@@ -747,14 +745,18 @@ internal static partial class EncodingTable
...
 */
+
+        // Protects the two maps below
+        static object mutex = new Object ();
+
         //
         // This caches the mapping of an encoding name to a code page.
         //
-        private static Hashtable hashByName = Hashtable.Synchronized(new Hashtable(StringComparer.OrdinalIgnoreCase));
+        private static Dictionary<string, int> hashByName = new Dictionary<string, int> (StringComparer.OrdinalIgnoreCase);
```
Comment (Therzok) [inline mcs/class/corlib/ReferenceSources/EncodingTable.cs]: The two dictionaries were independent before the refactoring. Better to lock on `hashByName.SyncRoot`?

Comment (vargaz) [reply mcs/class/corlib/ReferenceSources/EncodingTable.cs]: Fixed.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by lewing** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build OS X x64 Android SDK

**Comment by marek-safar** [comment]
note, similar change was rejected upstream due to performance impact.

**Pull request merged**